microbench/src/main/java/io/netty/microbench/channel/DefaultChannelPipelineBenchmark.java

```diff
@@ -55,30 +82,58 @@ public boolean isSharable() {
             return true;
         }
     };
+    private static final int CALL_TYPE_ARRAY_SIZE = 1024;
+    private static final int CALL_TYPE_ARRAY_MASK = CALL_TYPE_ARRAY_SIZE - 1;
 
     @Param({ "4" })
     public int extraHandlers;
 
     private ChannelPipeline pipeline;
 
+    private int[] callTypes;
+    private int callTypeCounter;
+
     @Setup(Level.Iteration)
     public void setup() {
         pipeline = new EmbeddedChannel().pipeline();
         for (int i = 0; i < extraHandlers; i++) {
             pipeline.addLast(NOOP_HANDLER);
         }
         pipeline.addLast(CONSUMING_HANDLER);
+        callTypes = new int[CALL_TYPE_ARRAY_SIZE];
+        SplittableRandom rng = new SplittableRandom();
+        for (int i = 0; i < callTypes.length; i++) {
+            callTypes[i] = rng.nextInt(0, 6);
+        }
     }
 
     @TearDown
     public void tearDown() {
         pipeline.channel().close();
     }
 
+    @CompilerControl(CompilerControl.Mode.DONT_INLINE)
     @Benchmark
     public void propagateEvent(Blackhole hole) {
-        for (int i = 0; i < 100; i++) {
-            hole.consume(pipeline.fireChannelReadComplete());
+        hole.consume(pipeline.fireChannelReadComplete());
+    }
+
+    @OperationsPerInvocation(16)
+    @CompilerControl(CompilerControl.Mode.DONT_INLINE)
+    @Benchmark()
+    public void propagateVariety(Blackhole hole) {
+        int i;
+        for (i = 0; i < 16; i++) {
+            int type = callTypes[callTypeCounter + i & CALL_TYPE_ARRAY_MASK];
+            switch (type) {
```

Comment:
This risks to be annoying for the branch predictor due to the switch - I would focus instead on making each propagation to use different pipelines which have slightly different propagation patterns e.g. outbound or inbound events which would populate the cache differently (if makes sense).

If you make the number of different pipelines configurable, when you exceed a certain number, with the current approach using the cache, you should see an unexpected degradation due to misprediction on the if used to find the next ctx node

Reply:
I'll also move the benchmark changes to a separate PR, to make it easier to compare the branches.